crates/revm/src/executor.rs

```diff
@@ -241,7 +241,7 @@ where
 
         #[cfg(feature = "optimism")]
         let l1_block_info =
-            self.chain_spec.optimism.then_some(optimism::L1BlockInfo::try_from(block)?);
+            self.chain_spec.optimism.then(|| optimism::L1BlockInfo::try_from(block)).transpose()?;
```

Comment:
tl;dr `.then_some` eagerly evaluates its params, causing the error to be thrown even if the chainspec doesn't have `optimism` enabled